src/routes/write-exercise/components/WriteExercise.tsx

```diff
@@ -130,23 +127,49 @@ const WriteExercise = ({
   const giveUp = async (): Promise<void> => {
     const previousResult = result
     setResult('incorrect')
-    storeResult(previousResult ?? 'incorrect')
+    storeResult(previousResult ?? 'incorrect', true)
   }
 
   const continueExercise = (): void => {
     setResult(null)
     setSubmission(null)
     setInput('')
 
-    if (currentDocumentNumber === documents.length - 1) {
+    if (currentDocumentNumber === documents.length - 1 && !getRetryInfoOfCurrent().needsToBeRepeated) {
       finishExercise(results)
     } else {
-      setCurrentDocumentNumber(currentDocumentNumber + 1)
+      getRetryInfoOfCurrent().needsToBeRepeated ? tryLater() : setCurrentDocumentNumber(currentDocumentNumber + 1)
     }
   }
 
-  const storeResult = (score: SimpleResultType): void => {
-    setResults([...results, { ...document, result: score }])
+  const storeResult = (score: SimpleResultType, noFurtherRetries: boolean = false): void => {
+    const nthRetry = noFurtherRetries ? 3 : getRetryInfoOfCurrent().nthRetry
+
+    const indexOfCurrentResult = results.findIndex(result => result.id === document?.id)
+    const result: DocumentResultType = { ...document, result: score, numberOfTries: nthRetry + 1 }
+    const newArr = results
+    if (indexOfCurrentResult !== -1) {
+      newArr[indexOfCurrentResult] = result
+      if (results[indexOfCurrentResult].result === 'similar') {
+        newArr[indexOfCurrentResult] = {
+          ...document,
+          result: nthRetry >= 2 ? 'incorrect' : 'similar',
+          numberOfTries: nthRetry + 1
+        }
+      }
+      setResult(newArr[indexOfCurrentResult].result)
+      setResults(newArr)
+    } else {
+      setResults([...results, result])
+    }
+  }
+
+  const getRetryInfoOfCurrent = (): { nthRetry: number; needsToBeRepeated: boolean } => {
+    const indexOfCurrent = results.findIndex(result => result.id === document?.id)
+    const nthRetry = indexOfCurrent === -1 ? 0 : results[indexOfCurrent].numberOfTries
+    const needsToBeRepeated =
+      nthRetry < 3 && (indexOfCurrent === -1 || results[indexOfCurrent].result !== SIMPLE_RESULTS.correct)
+    return { nthRetry, needsToBeRepeated }
```

Comment:
Same as in other exercise, refactor and execute once on top of file. Could you extract `3` to a constant?

Reply:
Moved 3 to a constant. Did not understand what you mean with the first part.

Reply:
This does not need to be part of a function. Just refactor using `find` as suggested in the other exercise and move it to the top.
I.e:
```
const { nthRetry, needsToBeRepeated } = results.find...
```

Reply:
will be done in LUN-219